OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -152,16 +152,24 @@ public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 			if (!mapBuildRadius.BuildRadiusEnabled)
 				return null;
 
-			foreach (var bp in world.ActorsWithTrait<BaseProvider>())
+			foreach (var a in world.ActorsHavingTrait<BaseProvider>())
 			{
-				var validOwner = bp.Actor.Owner == p || (allyBuildEnabled && bp.Actor.Owner.Stances[p] == Stance.Ally);
-				if (!validOwner || !bp.Trait.Ready())
-					continue;
-
-				// Range is counted from the center of the actor, not from each cell.
-				var target = Target.FromPos(bp.Actor.CenterPosition);
-				if (target.IsInRange(center, bp.Trait.Info.Range))
-					return bp.Actor;
+				var baseProviders = a.TraitsImplementing<BaseProvider>().ToArray();
```

Comment:
Isn't this just reimplementing a slower version of `ActorsWithTrait`?

Reply:
I'm not sure if i messed something else up or it should be like that, `ActorsWithTrait` was giving crash to me for multipile traits on the actor. So here finds the actors having trait and get all the traits from them.